app/uk/gov/hmrc/apisubscriptionfields/controller/ApiSubscriptionFieldsController.scala

```diff
@@ -17,25 +17,59 @@
+import play.api.mvc._
+import uk.gov.hmrc.apisubscriptionfields.model.ErrorCode._
+import uk.gov.hmrc.apisubscriptionfields.model._
+import uk.gov.hmrc.apisubscriptionfields.service.SubscriptionFieldsService
 
+import scala.concurrent.ExecutionContext.Implicits.global
 import scala.concurrent.Future
 
-trait ApiSubscriptionFieldsController extends BaseController {
-
-	def id() = Action.async { implicit request =>
-    /*
-     TODO:
-       parse `ApiSubscriptionRequest`
-       - [application_id]_[api_context]_[api_context] is the id of the mongo record, but decide good separators
-       - generate UUID (the header to be passed in the incoming requests)
-       - save record to mongo
-       */
-    val subscriptionFieldsId = UUID.randomUUID()
-    Future.successful(Created(s"$subscriptionFieldsId"))
+@Singleton
+class ApiSubscriptionFieldsController @Inject() (service: SubscriptionFieldsService) extends CommonController {
+
+  import JsonFormatters._
+
+  private val NOT_FOUND_RESPONSE = NotFound(JsErrorResponse(SUBSCRIPTION_FIELDS_ID_NOT_FOUND, "Subscription Fields were not found"))
+
+  def getSubscriptionFields(rawAppId: UUID, rawApiContext: String, rawApiVersion: String): Action[AnyContent] = Action.async { implicit request =>
+    val eventualMaybeResponse = service.get(SubscriptionIdentifier(AppId(rawAppId), ApiContext(rawApiContext), ApiVersion(rawApiVersion)))
+    asActionResult(eventualMaybeResponse)
+  }
+
+  def getSubscriptionFieldsByFieldsId(rawFieldsId: UUID): Action[AnyContent] = Action.async { implicit request =>
+    val eventualMaybeResponse = service.get(SubscriptionFieldsId(rawFieldsId))
+    asActionResult(eventualMaybeResponse)
   }
-}
 
-object ApiSubscriptionFieldsController extends ApiSubscriptionFieldsController
+  private def asActionResult(eventualMaybeResponse: Future[Option[SubscriptionFieldsResponse]]) = {
+    eventualMaybeResponse map {
+      case Some(subscriptionFields) => Ok(Json.toJson(subscriptionFields))
+      case None => NOT_FOUND_RESPONSE
+    } recover recovery
+  }
+
+  def upsertSubscriptionFields(rawAppId: UUID, rawApiContext: String, rawApiVersion: String): Action[JsValue] = Action.async(parse.json) { implicit request =>
+      withJsonBody[SubscriptionFieldsRequest] { payload =>
+        service.upsert(SubscriptionIdentifier(AppId(rawAppId), ApiContext(rawApiContext), ApiVersion(rawApiVersion)), payload.fields) map {
+          case (response, true) => Created(Json.toJson(response))
+          case (response, false) => Ok(Json.toJson(response))
+        } recover {
+          case e: Exception => BadRequest(e.getMessage)
+        }
+      } recover recovery
+    }
+
+  def deleteSubscriptionFields(rawAppId: UUID, rawApiContext: String, rawApiVersion: String): Action[AnyContent] = Action.async { implicit request =>
+    val identifier = SubscriptionIdentifier(AppId(rawAppId), ApiContext(rawApiContext), ApiVersion(rawApiVersion))
+    //TODO remove service.get
```

Comment:
I would replace this code with:
```
    service.delete(identifier) map {
      case true => NoContent
      case false => NOT_FOUND_RESPONSE
    } recover recovery
```

What do you think?

Reply:
done by @gokyo